src/python/grpcio_tests/tests_aio/unit/interceptor_test.py

```diff
@@ -0,0 +1,216 @@
+# Copyright 2019 The gRPC Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import asyncio
+import collections
+import logging
+import unittest
+
+import grpc
+
+from grpc.experimental import aio
+from tests_aio.unit._test_server import start_test_server
+from tests_aio.unit._test_base import AioTestBase
+from src.proto.grpc.testing import messages_pb2
+
+
+class _ClientCallDetails(
+        collections.namedtuple(
+            '_ClientCallDetails',
+            ('method', 'timeout', 'metadata', 'credentials')),
+        grpc.ClientCallDetails):
+    pass
```

Comment:
Named tuple doesn't have typing information. Is there any alternative?

Reply:
Considering that this is tests I won't be worried about typing since we are not enforncing typing in the tests, we want to do it?

Another discussion is how to improve this vague object [1] which might be not only an interface but also an implementation since the `ClientCallDetails` needs to be instantiated within the interceptors. For me there are two pending discussions here:

Do we need to make the `ClientCallDetails` a first citizen class implementation instead of having this vague interface since interceptors need to instantiate them? 

So having something like

```python
class ClientCallDetails(
        collections.namedtuple(
            '_ClientCallDetails',
            ('method', 'timeout', 'metadata', 'credentials')),
        grpc.ClientCallDetails):

    method: int
    timeout: float
    metadata: dict
    credentials: dict
```

We could even add a clone method for making life easier to the developers

```python
class ClientCallDetails(
        collections.namedtuple(
            '_ClientCallDetails',
            ('method', 'timeout', 'metadata', 'credentials')),
        grpc.ClientCallDetails):

    method: int
    timeout: float
    metadata: dict
    credentials: dict

    def clone(self, method=None, timeout=None, metadata=None, credentials=None):
        """"Clone the ClientDetails to a new one and override any of the attributes with a new
        value with giving a None value"""
```
The second discussion would be, can we make usage of something more explicit than a `namedtuple`?

- Yes a namedutple with some class typing?
- Use data classes [2] that they play well with typing. IIRC they are available since Python 3.7 so it won't be an option if I'm not wrong since we are providing support for 3.5.X and beyond.
- Use a regular Python class which again plays well with typing but might have some performance footprint.
- Use an external library like `attrs` [3] 

IMO I would go for an improved version fo the `namedtuple` by having thee class typing and providing a first citizen class implementation rather than an interface.

WDTY?

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/__init__.py#L406
[2] https://docs.python.org/3/library/dataclasses.html
[2] https://www.attrs.org/en/stable/

Reply:
I will go for making the `ClientCallDetails` as a first citizen class provided by the `Aio` module.

Reply:
First, if this is just an implementation of `grpc.ClientCallDetails`, maybe it shouldn't be a first class citizen.

Second, the `clone` method is reasonable. If we want to add it, we need to add it to both stack. Technically, adding new API should go through the [gRFC process](https://github.com/grpc/proposal). You can write a short document to give people a chance to comment, most gRFC are quite short.

Third, based on the [Google Style Guide](http://google.github.io/styleguide/pyguide.html#384-classes), the recommended way of writing class docstring is that: 1) short summary in one; 2) long summery can spread multiple lines, even include examples; 3) `Args`, `Returns`, `Exceptions` come after.